csharp/ql/src/experimental/Security Features/Serialization/DataSetSerialization.qll

```diff
@@ -0,0 +1,107 @@
+/**
+ * Provides classes for `DataSet` or `DataTable` deserialization queries.
+ *
+ * Please visit https://go.microsoft.com/fwlink/?linkid=2132227 for details.
+ */
+
+import csharp
+
+/**
+ * Abstract class that depends or inherits from `DataSet` or `DataTable` types.
+ */
+abstract class DataSetOrTableRelatedClass extends Class { }
+
+/**
+ * `DataSet`, `DataTable` types, or any types derived from them.
+ */
+class DataSetOrTable extends DataSetOrTableRelatedClass {
+  DataSetOrTable() {
+    this.getABaseType*().getQualifiedName() = "System.Data.DataTable" or
+    this.getABaseType*().getQualifiedName() = "System.Data.DataSet"
+  }
+}
+
+/**
+ * A Class that include a property or generic collection of type `DataSet` and `DataTable`
+ */
+class ClassWithDataSetOrTableMember extends DataSetOrTableRelatedClass {
+  ClassWithDataSetOrTableMember() {
+    this.getAMember().(AssignableMember).getType() instanceof DataSetOrTable
+    or
+    exists(Property p | p = this.getAProperty() |
+      p.getType() instanceof DataSetOrTable or
+      p.getType().(ConstructedGeneric).getATypeArgument() instanceof DataSetOrTable
+    )
+  }
+}
+
+/**
+ * Serializable types
+ */
+class SerializableClass extends Class {
+  SerializableClass() {
+    (
+      this.getABaseType*().getQualifiedName() = "System.Xml.Serialization.XmlSerializer" or
+      this.getABaseInterface*().getQualifiedName() = "System.Runtime.Serialization.ISerializable" or
+      this.getABaseType*().getQualifiedName() = "System.Runtime.Serialization.XmlObjectSerializer" or
+      this.getABaseInterface*().getQualifiedName() =
+        "System.Runtime.Serialization.ISerializationSurrogateProvider" or
+      this.getABaseType*().getQualifiedName() =
+        "System.Runtime.Serialization.XmlSerializableServices" or
+      this.getABaseInterface*().getQualifiedName() = "System.Xml.Serialization.IXmlSerializable"
+    )
+    or
+    exists(Attribute a | a = this.getAnAttribute() |
+      a.getType().getQualifiedName().toString() = "System.SerializableAttribute"
+    )
+  }
+}
+
+/**
+ * Holds if the serializable class `c` has a property or field `m` that is of `DataSet` or `DataTable` related type
+ */
+predicate isClassUnsafeXmlSerializerImplementation(SerializableClass c, Member m) {
+  exists(Property p | m = p |
+    p = c.getAProperty() and
+    p.getType() instanceof DataSetOrTableRelatedClass
+  )
+  or
+  exists(AssignableMember am | am = m |
+    (am = c.getAField() or am = c.getAMember()) and
+    am.getType() instanceof DataSetOrTableRelatedClass
+  )
+}
```

Comment:
This can be simplified to
```
predicate isClassUnsafeXmlSerializerImplementation(SerializableClass c, AssignableMember am) {
  am = c.getAMember() and
  am.getType() instanceof DataSetOrTableRelatedClass
}
```

Reply:
Fixing it

Reply:
Fixed both of them. But I still have no idea why the .qhelp file check is failing as I have no access to the details. Any advise will be appreciated.

Reply:
The qhelp check succeeds now. There are still 4 unaddressed comments:
https://github.com/github/codeql/pull/3951#discussion_r463461083
https://github.com/github/codeql/pull/3951#discussion_r463461135
https://github.com/github/codeql/pull/3951#discussion_r463463004
https://github.com/github/codeql/pull/3951#discussion_r463463218

Reply:
Fixed. Thanks, I missed those comments before.